smpclient/transport/ble.py

```diff
@@ -41,24 +41,51 @@ class SMPBLETransport:
     def __init__(self) -> None:
         self._buffer = bytearray()
         self._notify_condition = asyncio.Condition()
+        self.connected = False
         logger.debug(f"Initialized {self.__class__.__name__}")
 
+    async def _notify_disconnected(self):
+        async with self._notify_condition:
+            self.connected = False
+            self._notify_condition.notify()
+    
+    def _handle_disconnect(self, _):
+        if not self.connected:
+            return
+        logger.info(f"Disconnected")
+        asyncio.create_task(self._notify_disconnected())
+            
+    def _check_connected(self):
+        if not self.connected:
+            raise SMPBLETransportDeviceNotFound(f"Device is disconnected")
+
     async def connect(self, address: str) -> None:
         logger.debug(f"Scanning for {address=}")
-        device = (
-            await BleakScanner.find_device_by_address(address)  # type: ignore # upstream fix
-            if MAC_ADDRESS_PATTERN.match(address) or UUID_PATTERN.match(address)
-            else await BleakScanner.find_device_by_name(address)  # type: ignore # upstream fix
-        )
-
-        if type(device) is BLEDevice:
-            self._client = BleakClient(device, services=(str(SMP_SERVICE_UUID),))
+        if MAC_ADDRESS_PATTERN.match(address):
+            device = address
         else:
-            raise SMPBLETransportDeviceNotFound(f"Device '{address}' not found")
+            device = (
+                await BleakScanner.find_device_by_address(address)  # type: ignore # upstream fix
+                if UUID_PATTERN.match(address)
+                else await BleakScanner.find_device_by_name(address)  # type: ignore # upstream fix
+            )
+
+            if type(device) is not BLEDevice:
+                raise SMPBLETransportDeviceNotFound(f"Device '{address}' not found")
+            logger.debug(f"Found device: {device=}, connecting...")
+
+        self._client = BleakClient(device, services=(str(SMP_SERVICE_UUID),),
+                                   disconnected_callback=self._handle_disconnect)
 
-        logger.debug(f"Found device: {device=}, connecting...")
         await self._client.connect()
         logger.debug(f"Connected to {device=}")
+        self.connected = True
+        
+        # BlueZ doesn't have a proper way to get the MTU, so we have this hack.
+        # If this doesn't work for you, you can set the client._mtu_size attribute
+        # to override the value instead.
+        if self._client._backend.__class__.__name__ == "BleakClientBlueZDBus":
+            await self._client._backend._acquire_mtu()
 
```

Comment:
It sometimes returns incorrect value. It seems to be a bug somewhere in bleak.

Reply:
I'm using your project for https://github.com/esphome/esphome/pull/6075. It is quite convince to have all transport inside client since I don't have to include smpmgr.

Reply:
Oh cool!  I used to work on/play with Home Asssistant!  NRF52840 should have much better quiescent current than ESP32 if I understand correctly.  Wires/battery tradeoffs are always present.  Using feather boards or what?

Reply:
I have that one https://github.com/joric/nrfmicro/wiki/Alternatives#supermini-nrf52840. In sleep it takes around 200uA. You could get lower if you change LDO.

Reply:
GDI https://github.com/hbldh/bleak/issues/1497